File: src/data_rover/core/validation/rules/reach.py
"""Static reverse-reach: which elements can a mutation's dirty set affect
through user rules?

Each relationship atom in a rule contributes one root-first path of
(rel-type closure, direction, far-type closure) steps. Expansion walks
dirty elements backwards along every path suffix they could sit on and
adds the reached applies_to-typed elements. Over-approximation is safe
(the scoped rerun just revalidates a few extra elements); too-small is
the only correctness hazard.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

from .schema import (
    AllCond,
    AnyCond,
    Condition,
    NotCond,
    RelationshipCond,
    Rule,
)

if TYPE_CHECKING:
    from ...metamodel.schema import Metamodel
    from ...model.model import Model
    from .compile import CompiledRules


@dataclass(frozen=True)
class ReverseStep:
    rel_types: frozenset[str]
    #: direction as written in the rule: owner -> far
    direction: Literal["outgoing", "incoming"]
    far_types: frozenset[str] | None


@dataclass(frozen=True)
class ReversePath:
    steps: tuple[ReverseStep, ...]  # root-first


def derive_paths(rule: Rule, metamodel: Metamodel) -> list[ReversePath]:
    paths: list[ReversePath] = []

    def walk(cond: Condition, prefix: tuple[ReverseStep, ...]) -> None:
        match cond:
            case AllCond(all=subs) | AnyCond(any=subs):
                for c in subs:
                    walk(c, prefix)
            case NotCond(not_=sub):
                walk(sub, prefix)
            case RelationshipCond(relationship=spec):
                step = ReverseStep(
                    rel_types=metamodel.relationship_descendants(spec.type),
                    direction=spec.direction,
                    far_types=(
                        None
                        if spec.to is None
                        else metamodel.element_descendants(spec.to)
                    ),
                )
                paths.append(ReversePath(steps=(*prefix, step)))
                if spec.where is not None:
                    walk(spec.where, (*prefix, step))
            case _:
                pass  # property atoms reach only the element itself

    for cond in (rule.when, rule.then):
        if cond is not None:
            walk(cond, ())
    return paths


def _owners(model: Model, frontier: set[str], step: ReverseStep) -> set[str]:
    """One reverse hop: the elements whose `step` hop reaches `frontier`."""
    out: set[str] = set()
    for eid in frontier:
        rel_ids = (
            model.indexes.incoming_ids(eid)
            if step.direction == "outgoing"
            else model.indexes.outgoing_ids(eid)
        )
        for rid in rel_ids:
            rel = model.relationships[rid]
            if rel.type_name not in step.rel_types:
                continue
            out.add(rel.source_id if step.direction == "outgoing" else rel.target_id)
    return out


def expand_scope(
    model: Model, compiled: CompiledRules, dirty_ids: Iterable[str]
) -> list[str]:
    if not compiled.rules:
        return []
    extra: dict[str, None] = {}
    dirty_elements = [
        el for eid in dirty_ids if (el := model.elements.get(eid)) is not None
    ]
    for cr in compiled.rules:
        for path in cr.paths:
            steps = path.steps
            for d in range(1, len(steps) + 1):
                far = steps[d - 1].far_types
                seeds = {
                    el.id for el in dirty_elements if far is None or el.type_name in far
                }
                if not seeds:
                    continue
                frontier = seeds
                for i in range(d - 1, -1, -1):
                    frontier = _owners(model, frontier, steps[i])
                    gate = steps[i - 1].far_types if i > 0 else cr.applies_types
                    if gate is not None:
                        frontier = {
                            fid
                            for fid in frontier
                            if (fel := model.elements.get(fid)) is not None
                            and fel.type_name in gate
                        }
                    if not frontier:
                        break
                for fid in sorted(frontier):
                    extra[fid] = None
    return list(extra)

File: engine/src/rules/reach.ts
import type { Metamodel } from '../metamodel/metamodel.ts';
import type { Model } from '../model/model.ts';
import type { ElementRec } from '../model/records.ts';
import { cmpCodePoint } from '../value/compare.ts';
import type { CompiledRules } from './compile.ts';
import type { Condition, Rule } from './document.ts';

/** One hop of a rule's path, as written: owner to far element. */
export type ReverseStep = {
	relTypes: ReadonlySet<string>;
	direction: 'outgoing' | 'incoming';
	farTypes: ReadonlySet<string> | null;
};

/** The hops from a rule's owner to the far element of one relationship atom, root first. */
export type ReversePath = readonly ReverseStep[];

/** A path per relationship atom of `when` then `then`; a `where` extends its atom's path. */
export function derivePaths(rule: Rule, mm: Metamodel): ReversePath[] {
	const paths: ReversePath[] = [];
	const walk = (cond: Condition, prefix: ReversePath): void => {
		if ('all' in cond) for (const sub of cond.all) walk(sub, prefix);
		else if ('any' in cond) for (const sub of cond.any) walk(sub, prefix);
		else if ('not' in cond) walk(cond.not, prefix);
		else if (!('property' in cond)) {
			const step: ReverseStep = {
				relTypes: mm.relationshipDescendants(cond.type),
				direction: cond.direction,
				farTypes: cond.to === null ? null : mm.elementDescendants(cond.to)
			};
			paths.push([...prefix, step]);
			if (cond.where !== null) walk(cond.where, [...prefix, step]);
		}
	};
	for (const cond of [rule.when, rule.then]) if (cond !== null) walk(cond, []);
	return paths;
}

/** One hop backwards: the elements whose `step` reaches `frontier`. */
function owners(frontier: ReadonlySet<ElementRec>, step: ReverseStep): Set<ElementRec> {
	const out = new Set<ElementRec>();
	for (const el of frontier) {
		if (step.direction === 'outgoing') {
			for (const rel of el.in) if (step.relTypes.has(rel.typeName)) out.add(rel.source);
		} else {
			for (const rel of el.out) if (step.relTypes.has(rel.typeName)) out.add(rel.target);
		}
	}
	return out;
}

const byId = (a: ElementRec, b: ElementRec) => cmpCodePoint(a.id, b.id);

/**
 * The elements whose rule verdicts a change to `dirtyIds` may move: from each
 * dirty element (relationship and deleted ids drop out), every suffix of every
 * rule path is walked backwards, and each element reached whose type the rule
 * applies to is kept, in first-seen order. The far types are never used to
 * filter: a retype on the path must not hide an owner.
 */
export function expandScope(
	model: Model,
	compiled: CompiledRules,
	dirtyIds: Iterable<string>
): string[] {
	if (compiled.rules.length === 0) return [];
	const seeds = new Set<ElementRec>();
	for (const id of dirtyIds) {
		const el = model.findElement(id);
		if (el !== undefined) seeds.add(el);
	}
	if (seeds.size === 0) return [];
	const extra = new Set<string>();
	for (const cr of compiled.rules) {
		for (const steps of cr.paths) {
			for (let d = 1; d <= steps.length; d++) {
				let frontier: ReadonlySet<ElementRec> = seeds;
				for (let i = d - 1; i >= 0; i--) {
					frontier = owners(frontier, steps[i]!);
					if (frontier.size === 0) break;
				}
				for (const el of [...frontier].sort(byId)) {
					if (cr.appliesTypes.has(el.typeName)) extra.add(el.id);
				}
			}
		}
	}
	return [...extra];
}
